### PR TITLE
Experimental BTF support

### PR DESCRIPTION
**Pull request by vadorovsky**
Provide a CLI option `--btf` which provides an experimental BTF support.

This includes a sanitization logic, which modifies the debug info in order to make it acceptable by the kernel, which includes:

* A requirement for names to be compatible with C.
* Removing DI for data-carrying enums.
* Removing name from types annotated with `AyaBtfMapMarker` - Linux kernel requires maps to be anonymous structures, but Rust doesn't support anonymous types. This is our way to make it working.
* Sorting type members by offsets (sometimes they arrive out of order).
* Setting linkage of all functions which are not BPF programs to `static`.

Updates: #1
Closes: #177 #26

### CODE REVIEW

File: src/llvm/types/ir.rs [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+    }
+
+    pub(crate) fn sub_program(&self, context: LLVMContextRef) -> Option<DISubprogram<'ctx>> {
+        let sub_program = unsafe { LLVMGetSubprogram(self.value_ref) };
+        NonNull::new(sub_program).map(|_| unsafe {
+            DISubprogram::from_value_ref(LLVMMetadataAsValue(context, sub_program))
+        })
+    }
+
+    pub(crate) fn set_subprogram(&mut self, sub_program: &DISubprogram) {
```
Comment (alessandrod) [inline src/llvm/types/ir.rs:336]: this should be set_sub_program

Comment (vadorovsky) [reply src/llvm/types/ir.rs:336]: I think "subprogram" is usually written as one word together, so probably the other way around?

https://llvm.org/doxygen/classllvm_1_1DISubprogram.html
https://www.oed.com/dictionary/subprogramme_n?tab=factsheet#20089009

**Review by dave-tucker — CHANGES_REQUESTED** [review]
A couple of things:

1. Could you take a pass through the open issues and make sure to add them to the `Closes: ` part of the PR description that I've added if they are fixed in this PR
2. Where are the docs? 😄 

A section in the README about "BTF Support" would be excellent, explaining how to pass the linker flag to bpf-linker when working on an Aya project.

File: src/linker.rs [hunk 2]
```diff
@@ -446,12 +464,19 @@ impl Linker {
                 &self.options.export_symbols,
             )
         }
-        .map_err(LinkerError::OptimizeError)
+        .map_err(LinkerError::OptimizeError)?;
+
+        // if we don't need BTF emission, we can strip DI
+        if !self.options.btf {
```
[the remark this answers is not in the document]
Comment (vadorovsky) [reply src/linker.rs]: Done

File: src/llvm/types/ir.rs (hunk 1 above)
Comment (vadorovsky) [reply src/llvm/types/ir.rs]: I changed to "subprogram", see the commit message

File: src/llvm/di.rs [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+            // There's no way to set the unit with LLVMDIBuilderCreateFunction
+            // so we set it after creation.
+            if let Some(unit) = sub_program.unit() {
+                new_program.set_unit(unit);
+            }
+
+            // Add retained nodes from the old program. This is needed to preserve local debug
+            // variables, including function arguments which otherwise become "anon". See
+            // LLVMDIBuilderFinalizeSubprogram and
+            // DISubprogram::replaceRetainedNodes.
```
[the remark this answers is not in the document]
Comment (vadorovsky) [reply src/llvm/di.rs]: Fixed

**Comment by vadorovsky** [comment]
@dave-tucker I added a small section in README. Do you think it's enough? I would rather document more details (like C-shims) in the book.

**Review by dave-tucker — APPROVED** [review]
Given most of these changes were reviewed via PR into the feature branch and a cursory glance over the newer commits, this LGTM.

File: src/linker.rs [hunk 4]
```diff
@@ -246,7 +249,22 @@ impl Linker {
         self.llvm_init();
         self.link_modules()?;
         self.create_target_machine()?;
+        if let Some(path) = &self.options.dump_module {
+            std::fs::create_dir_all(path).map_err(|err| LinkerError::IoError(path.clone(), err))?;
+        }
+        if let Some(path) = &self.options.dump_module {
```
Comment (alessandrod) [inline src/linker.rs:255]: This should be merged with the if above

**Review by alessandrod — COMMENTED** [review]
just the if thing, then can be merged

**Review by alessandrod — APPROVED** [review]

**Pull request merged**